evm/p2p/state.py

```diff
@@ -43,28 +46,39 @@ class StateDownloader:
     # loop.
     _last_report_time = 0
 
-    def __init__(self, chaindb, state_db, root_hash, network_id, privkey):
-        self.peer_pool = PeerPool(ETHPeer, chaindb, network_id, privkey, self.msg_handler)
+    def __init__(self, state_db: BaseDB, root_hash: bytes, peer_pool: PeerPool) -> None:
+        self.peer_pool = peer_pool
+        self.peer_pool.subscribe(self)
         self.root_hash = root_hash
         self.scheduler = StateSync(root_hash, state_db, self.logger)
-
-    def msg_handler(self, peer: BasePeer, cmd: protocol.Command,
-                    msg: protocol._DecodedMsgType) -> None:
-        """The callback passed to BasePeer, called for every incoming message."""
-        peer = cast(ETHPeer, peer)
-        if isinstance(cmd, eth.NodeData):
-            self.logger.debug("Processing NodeData with %d entries" % len(msg))
-            for node in msg:
-                self._total_processed_nodes += 1
-                node_key = keccak(node)
-                try:
-                    self.scheduler.process([(node_key, node)])
-                except SyncRequestAlreadyProcessed:
-                    # This means we received a node more than once, which can happen when we retry
-                    # after a timeout.
-                    pass
-                # A node may be received more than once, so pop() with a default value.
-                self._pending_nodes.pop(node_key, None)
+        self._running_peers = set()  # type: Set[ETHPeer]
+        self._should_stop = asyncio.Event()
+
+    def register_peer(self, peer: BasePeer) -> None:
+        asyncio.ensure_future(self.handle_peer(cast(ETHPeer, peer)))
+
+    async def handle_peer(self, peer: ETHPeer) -> None:
+        """Handle the lifecycle of the given peer."""
+        self._running_peers.add(peer)
+        while not self._should_stop.is_set():
+            try:
+                cmd, msg = await peer.read_sub_proto_msg()
+            except PeerFinished:
+                break
+            if isinstance(cmd, eth.NodeData):
+                self.logger.debug("Processing NodeData with %d entries" % len(msg))
+                for node in msg:
+                    self._total_processed_nodes += 1
+                    node_key = keccak(node)
+                    try:
+                        self.scheduler.process([(node_key, node)])
+                    except SyncRequestAlreadyProcessed:
+                        # This means we received a node more than once, which can happen when we
+                        # retry after a timeout.
+                        pass
+                    # A node may be received more than once, so pop() with a default value.
+                    self._pending_nodes.pop(node_key, None)
+        self._running_peers.remove(peer)
```

Comment:
Just curious if it makes sense for this to be embedded in a `try/finally`.  Not sure how critical it is that this get removed from the `_running_peers` in the case of an unforseen error?

Reply:
Yeah, it definitely makes sense. Fixed